src/SubHask/Algebra/Accelerate/Vector.hs

```diff
@@ -43,183 +37,198 @@ import Unsafe.Coerce
+instance (KnownNat n, Monoid r, ValidACCVector bknd n r) => Monoid (ACCVector (bknd::Backend) (n::Nat) r) where
+--     {-# INLINE zero #-}
+--     zero = ACCVector(A.use (A.fromList (A.Z A.:.1) [(0::r)]))
 
-instance (Group r, ValidACCVector bknd n r) => Group (ACCVector (bknd::Backend) (n::Symbol) r) where
+instance (KnownNat n, Group r, ValidACCVector bknd n r) => Group (ACCVector (bknd::Backend) (n::Nat) r) where
     {-# INLINE negate #-}
     negate = negate
 
-instance (Monoid r, Abelian r, ValidACCVector bknd n r) => Abelian (ACCVector (bknd::Backend)  (n::Symbol) r)
+instance (KnownNat n, Monoid r, Abelian r, ValidACCVector bknd n r) => Abelian (ACCVector (bknd::Backend)  (n::Nat) r)
 
-instance (FreeModule r, ValidACCVector bknd n r, IsScalar r) => FreeModule (ACCVector (bknd::Backend)  (n::Symbol) r) where
+instance (KnownNat n, FreeModule r, ValidACCVector bknd n r, IsScalar r) => FreeModule (ACCVector (bknd::Backend)  (n::Nat) r) where
     {-# INLINE (.*.)   #-}
     (.*.) (ACCVector a1) (ACCVector a2) = ACCVector( A.zipWith (P.*) a1 a2)
 
-instance (Module r, ValidACCVector bknd n r, IsScalar r) => Module (ACCVector (bknd::Backend) (n::Symbol) r) where
+instance (KnownNat n, Module r, ValidACCVector bknd n r, IsScalar r) => Module (ACCVector (bknd::Backend) (n::Nat) r) where
     {-# INLINE (.*)   #-}
-    (.*) (ACCVector  v) r = ACCVector (A.map (P.* (A.constant r)) v)
+    (.*) (ACCVector  v) r = ACCVector (A.map (P.* r) v)
 
-instance (VectorSpace r, ValidACCVector bknd n r, IsScalar r) => VectorSpace (ACCVector (bknd::Backend) (n::Symbol) r) where
+instance (KnownNat n, VectorSpace r, ValidACCVector bknd n r, IsScalar r) => VectorSpace (ACCVector (bknd::Backend) (n::Nat) r) where
     {-# INLINE (./)   #-}
-    (./) (ACCVector  v) r = ACCVector (A.map (P./ (A.constant r)) v)
+    (./) (ACCVector  v) r = ACCVector (A.map (P./ r) v)
 
     {-# INLINE (./.)  #-}
     (./.) (ACCVector a1) (ACCVector a2) = ACCVector (A.zipWith (P./) a1 a2)
 
+-- Could not deduce (r ~ Elem r)
+-- In the instance declaration for ‘FiniteModule (ACCVector b n r)’
 
---Full error from FiniteModule instance:
-  -- Could not deduce (r ~ A.Exp r)
-  -- from the context (FreeModule r,
-  --                   ValidLogic r,
-  --                   ValidACCVector b n r,
-  --                   IsScalar r)
-  --   bound by the instance declaration
-  --   at src/SubHask/Algebra/Accelerate/Vector.hs:123:10-115
-  --   ‘r’ is a rigid type variable bound by
-  --       the instance declaration
-  --       at src/SubHask/Algebra/Accelerate/Vector.hs:123:10
-  -- In the instance declaration for ‘FiniteModule (ACCVector b n r)’
-
-
-instance (FreeModule r, ValidLogic r, ValidACCVector b n r, IsScalar r) => FiniteModule (A.Exp (ACCVector b (n::Symbol) r))
+instance (KnownNat n, FreeModule r, ValidLogic r, ValidACCVector b n r, IsScalar r) => FiniteModule (ACCVector b (n::Nat) r)
```

Comment:
Is there something defined about FiniteModule such that it requires an `Int`?  This instance (along with the `dim` function below) both yield `A.Exp Int` which is the Acclerate scalar int datatype, but I can't seem to make either type check since A.Exp Int is not an `Int.`